pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/BlockAwareSegmentInputStream.java

```diff
@@ -90,68 +84,60 @@ private int readEntries() throws IOException {
+            if ((!entriesByteBuf.isEmpty()) && bytesReadOffset + entriesByteBuf.get(0).readableBytes() > blockSize) {
+                // not able to place a new Entry.
+                entriesByteBuf.forEach(buf -> buf.release());
+                entriesByteBuf.clear();
+                dataBlockFullOffset = bytesReadOffset;
             }
-
-            return ret;
-        } catch (InterruptedException | ExecutionException e) {
-            log.error("Exception when get CompletableFuture<LedgerEntries>. ", e);
-            throw new IOException(e);
         }
-    }
 
-    // read entries from ledger, and pre-handle the returned ledgerEntries.
-    private void readNextEntriesFromLedger() throws InterruptedException, ExecutionException {
-        checkState(bytesReadOffset == blockBytesHave);
+        return ret;
+    }
 
-        long start = startEntryId + blockEntryCount;
-        long end = Math.min(start + entriesNumberEachRead - 1, ledger.getLastAddConfirmed());
+    private List<ByteBuf> readNextEntriesFromLedger(long start, long maxNumberEntries) throws IOException {
+        long end = Math.min(start + maxNumberEntries - 1, ledger.getLastAddConfirmed());
         try (LedgerEntries ledgerEntriesOnce = ledger.readAsync(start, end).get()) {
             log.debug("read ledger entries. start: {}, end: {}", start, end);
 
+            List<ByteBuf> entries = Lists.newLinkedList();
+
             Iterator<LedgerEntry> iterator = ledgerEntriesOnce.iterator();
-            long entryId = start;
             while (iterator.hasNext()) {
                 LedgerEntry entry = iterator.next();
                 int entryLength = (int) entry.getLength();
-                entryId = entry.getEntryId();
-
-                if (blockSize - blockBytesHave >= entryLength + entryHeaderSize) {
-                    // data block has space for this entry, keep this entry
-                    CompositeByteBuf entryBuf = PooledByteBufAllocator.DEFAULT.compositeBuffer();
-                    ByteBuf entryHeaderBuf = PooledByteBufAllocator.DEFAULT.buffer(entryHeaderSize, entryHeaderSize);
-
-                    entryHeaderBuf.writeInt(entryLength).writeLong(entryId);
-                    entryBuf.addComponents(entryHeaderBuf, entry.getEntryBuffer().retain());
-                    entryBuf.writerIndex(entryHeaderSize + entryLength);
-
-                    entriesByteBuf.add(entryBuf);
-
-                    // set counters
-                    blockEntryCount++;
-                    payloadBytesHave += entryLength;
-                    blockBytesHave += entryLength + entryHeaderSize;
-                } else {
-                    // data block has no space left for a whole message entry
-                    dataBlockFullOffset = blockBytesHave;
-                    break;
-                }
+                long entryId = entry.getEntryId();
+
+                CompositeByteBuf entryBuf = PooledByteBufAllocator.DEFAULT.compositeBuffer();
```

Comment:
nit: specify that the compositeBuffer will have 2 buffers since, we know it will.